File: packages/nemoclaw-haystack-agent/config/generate-config.ts
// SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

// Generate the credential-free Fabric projection for the local Haystack POC.

import { chmodSync, mkdirSync, writeFileSync } from "node:fs";
import { homedir } from "node:os";
import { join } from "node:path";

const SUPPORTED_INFERENCE_API = "openai-completions";
const MANAGED_BASE_URL = "https://inference.local/v1";
const FABRIC_ADAPTER_PATH = "/usr/local/share/nemoclaw/haystack-agent.fabric-adapter.json";
const FABRIC_ARTIFACTS_PATH = "/sandbox/.haystack-agent/fabric-artifacts";
const FABRIC_API_KEY_ENV = "HAYSTACK_FABRIC_API_KEY";
const DEFAULT_TEMPERATURE = 0;
const DEFAULT_MAX_AGENT_STEPS = 8;
const MAX_AGENT_STEPS = 32;

function normalizeText(value: string | undefined, name: string): string {
  const text = (value ?? "").trim();
  if (!text) throw new Error(`${name} is required.`);
  if (/[\p{Cc}\p{Cf}]/u.test(text)) {
    throw new Error(`${name} must not contain control characters.`);
  }
  return text;
}

function normalizeInferenceApi(value: string | undefined): string {
  const api = normalizeText(value ?? SUPPORTED_INFERENCE_API, "NEMOCLAW_INFERENCE_API");
  if (api !== SUPPORTED_INFERENCE_API) {
    throw new Error(
      `NEMOCLAW_INFERENCE_API must be ${SUPPORTED_INFERENCE_API} for Haystack Agent.`,
    );
  }
  return api;
}

function normalizeBaseUrl(value: string | undefined): string {
  const baseUrl = normalizeText(value ?? MANAGED_BASE_URL, "NEMOCLAW_INFERENCE_BASE_URL");
  let parsed: URL;
  try {
    parsed = new URL(baseUrl);
  } catch {
    throw new Error("NEMOCLAW_INFERENCE_BASE_URL must be a valid URL.");
  }
  if (parsed.username || parsed.password) {
    throw new Error("NEMOCLAW_INFERENCE_BASE_URL must not include credentials.");
  }
  if (baseUrl !== MANAGED_BASE_URL) {
    throw new Error(`NEMOCLAW_INFERENCE_BASE_URL must be ${MANAGED_BASE_URL}.`);
  }
  return baseUrl;
}

function normalizeTemperature(value: string | undefined): number {
  if (value === undefined || value.trim() === "") return DEFAULT_TEMPERATURE;
  const temperature = Number(value);
  if (!Number.isFinite(temperature) || temperature < 0 || temperature > 2) {
    throw new Error("NEMOCLAW_TEMPERATURE must be a number between 0 and 2.");
  }
  return temperature;
}

function normalizeMaxSteps(value: string | undefined): number {
  if (value === undefined || value.trim() === "") return DEFAULT_MAX_AGENT_STEPS;
  if (!/^\d+$/u.test(value.trim())) {
    throw new Error(`NEMOCLAW_MAX_TURNS must be an integer between 1 and ${MAX_AGENT_STEPS}.`);
  }
  const maxSteps = Number(value);
  if (!Number.isSafeInteger(maxSteps) || maxSteps < 1 || maxSteps > MAX_AGENT_STEPS) {
    throw new Error(`NEMOCLAW_MAX_TURNS must be an integer between 1 and ${MAX_AGENT_STEPS}.`);
  }
  return maxSteps;
}

function buildFabricConfig(env: NodeJS.ProcessEnv): Readonly<Record<string, unknown>> {
  const model = normalizeText(env.NEMOCLAW_MODEL, "NEMOCLAW_MODEL");
  const baseUrl = normalizeBaseUrl(env.NEMOCLAW_INFERENCE_BASE_URL);
  normalizeInferenceApi(env.NEMOCLAW_INFERENCE_API);
  const systemInstruction = normalizeText(
    env.NEMOCLAW_SYSTEM_INSTRUCTION ??
      "Answer the user's request directly. Do not claim tools or durable memory.",
    "NEMOCLAW_SYSTEM_INSTRUCTION",
  );
  return {
    schema_version: "fabric.agent/v1alpha1",
    metadata: {
      name: "nemoclaw-haystack-agent",
      description: "NemoClaw local Haystack Agent POC",
    },
    harness: {
      adapter_id: "nvidia.nemoclaw.haystack-agent",
      resolution: "preinstalled",
    },
    discovery: {
      local_paths: [FABRIC_ADAPTER_PATH],
    },
    instructions: {
      system: { content: systemInstruction, mode: "replace" },
    },
    runtime: {
      input_schema: "text",
      output_schema: "message",
      artifacts: FABRIC_ARTIFACTS_PATH,
      timeout_seconds: 90,
      max_turns: normalizeMaxSteps(env.NEMOCLAW_MAX_TURNS),
    },
    environment: {
      provider: "local",
      workspace: "/sandbox",
      artifacts: FABRIC_ARTIFACTS_PATH,
      ownership: "caller_owned",
      control_location: "in_env_control",
    },
    models: {
      default: {
        provider: "openshell",
        model,
        api_key_env: FABRIC_API_KEY_ENV,
        temperature: normalizeTemperature(env.NEMOCLAW_TEMPERATURE),
        base_url: baseUrl,
      },
    },
  };
}

function main(): void {
  const configDirectory = join(homedir(), ".haystack-agent");
  const configPath = join(configDirectory, "fabric.json");
  mkdirSync(configDirectory, { recursive: true, mode: 0o700 });
  writeFileSync(configPath, `${JSON.stringify(buildFabricConfig(process.env), null, 2)}\n`, {
    mode: 0o600,
  });
  chmodSync(configPath, 0o600);
  console.log(`[config] Wrote ${configPath}.`);
}

main();

File: packages/nemoclaw-deepseek-harness/config/generate-config.py
# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0

"""Generate the credential-free Fabric configuration for DeepSeek Harness."""

from __future__ import annotations

import json
import os
from pathlib import Path
from urllib.parse import urlsplit


ADAPTER_ID = "nvidia.nemoclaw.deepseek-harness"
ADAPTER_PATH = "/usr/local/share/nemoclaw/deepseek.fabric-adapter.json"
API_KEY_ENV = "DEEPSEEK_FABRIC_API_KEY"
ARTIFACTS_PATH = "/sandbox/.deepseek-harness/fabric-artifacts"
BASE_URL = "https://inference.local/v1"
INFERENCE_API = "openai-completions"


def _normalized_text(value: str | None, name: str) -> str:
    """Return one non-empty metadata value without invisible controls."""

    if value is None:
        raise ValueError(f"{name} is required")
    text = value.strip()
    if not text or any(
        ord(character) < 32 or ord(character) == 127 for character in text
    ):
        raise ValueError(f"{name} must be non-empty text without control characters")
    return text


def _managed_base_url(value: str | None) -> str:
    """Accept only the OpenShell-managed inference route used by the adapter."""

    text = _normalized_text(value or BASE_URL, "NEMOCLAW_INFERENCE_BASE_URL")
    parsed = urlsplit(text)
    if parsed.username or parsed.password or parsed.query or parsed.fragment:
        raise ValueError(
            "NEMOCLAW_INFERENCE_BASE_URL must not contain credentials or URL metadata"
        )
    if text != BASE_URL:
        raise ValueError(f"NEMOCLAW_INFERENCE_BASE_URL must be {BASE_URL}")
    return text


def build_fabric_config(environment: dict[str, str]) -> dict[str, object]:
    """Project NemoClaw build settings onto the released Fabric schema."""

    model = _normalized_text(environment.get("NEMOCLAW_MODEL"), "NEMOCLAW_MODEL")
    inference_api = _normalized_text(
        environment.get("NEMOCLAW_INFERENCE_API", INFERENCE_API),
        "NEMOCLAW_INFERENCE_API",
    )
    if inference_api != INFERENCE_API:
        raise ValueError(f"NEMOCLAW_INFERENCE_API must be {INFERENCE_API}")
    base_url = _managed_base_url(environment.get("NEMOCLAW_INFERENCE_BASE_URL"))
    return {
        "schema_version": "fabric.agent/v1alpha1",
        "metadata": {
            "name": "nemoclaw-deepseek-harness",
            "description": "NemoClaw-managed DeepSeek Harness headless runtime",
        },
        "harness": {"adapter_id": ADAPTER_ID, "resolution": "preinstalled"},
        "discovery": {"local_paths": [ADAPTER_PATH]},
        "runtime": {
            "input_schema": "text",
            "output_schema": "message",
            "artifacts": ARTIFACTS_PATH,
            "timeout_seconds": 90,
        },
        "environment": {
            "provider": "local",
            "workspace": "/sandbox",
            "artifacts": ARTIFACTS_PATH,
            "ownership": "caller_owned",
            "control_location": "in_env_control",
        },
        "models": {
            "default": {
                "provider": "openshell",
                "model": model,
                "api_key_env": API_KEY_ENV,
                "base_url": base_url,
            }
        },
    }


def write_fabric_config(environment: dict[str, str]) -> Path:
    """Write private configuration and state directories beneath explicit DSH_HOME."""

    home = Path(environment.get("HOME", "/sandbox"))
    state_root = home / ".deepseek-harness"
    state_root.mkdir(parents=True, exist_ok=True, mode=0o700)
    state_root.chmod(0o700)
    for directory in ("sessions", "logs", "cache", "fabric-artifacts"):
        child = state_root / directory
        child.mkdir(exist_ok=True, mode=0o700)
        child.chmod(0o700)
    config_path = state_root / "fabric.json"
    config_path.write_text(
        json.dumps(build_fabric_config(environment), indent=2) + "\n",
        encoding="utf-8",
    )
    config_path.chmod(0o600)
    return config_path


def main() -> None:
    """Generate the package configuration used by the terminal command."""

    path = write_fabric_config(dict(os.environ))
    print(f"[config] Wrote {path}")


if __name__ == "__main__":
    main()
